File: api/services/reply.py
from services.pg import get_db_connection, fetch_relevant_posts
from services.llm import get_embedding, search_used_items
from services.messages import format_telegram_message

async def get_marketplace_reply(message_data: dict):

  user_query = message_data.get("text", "").strip()
  if not user_query:
    return "⚠️ What used item are you looking for today?"

  # 🔍 Embedding
  query_vector = await get_embedding(user_query)
  if not query_vector:
    return "❌ Sorry, I had trouble processing your search. Please try again."

  # 🗄️ Fetch from DB
  try:
    with get_db_connection() as conn:
      relevant_posts = fetch_relevant_posts(conn, query_vector, limit=5)
  except Exception as e:
    print(f"DB Error: {e}")
    return "❌ Database connection issue. Please try again later."

  if not relevant_posts:
    return "🔍 No matching items found. Try a different keyword!"

  # 🧠 LLM Analysis
  result = await search_used_items(user_query, relevant_posts)
  if not result or not result.listings:
    return "❌ Failed to analyze results. Try a more specific search."

  # 🧾 Format final Telegram message
  try:
    reply = format_telegram_message(
      user_query=user_query,
      result=result,
      relevant_posts=relevant_posts
    )
    return reply

  except Exception as e:
    print(f"Format Error: {e}")
    return "❌ Error formatting results. Please try again."

File: internal/bot-server/reply.go
package botserver

import (
  "context"
  "log"
  "strings"
  "time"

  "leaddits/internal/pkg"
)

// GetMarketplaceReply handles the full pipeline: Embedding -> DB -> LLM -> Formatting
func GetMarketplaceReply(ctx context.Context, text string) string {
  userQuery := strings.TrimSpace(text)
  if userQuery == "" {
    return "⚠️ What used item are you looking for today?"
  }

  // 1. Initialize DB Connection (Usually passed in or managed via a singleton)
  db, err := pkg.Connect(GlobalConfig.ConnStr)
  if err != nil {
    log.Printf("DB Connection Error: %v", err)
    return "❌ Database connection issue. Please try again later."
  }
  defer db.Close()

  // 2. 🔍 Get Embedding
  // We wrap this in a timeout to prevent long hangs
  embedCtx, cancelEmbed := context.WithTimeout(ctx, 10*time.Second)
  defer cancelEmbed()

  queryVector, err := GetEmbedding(embedCtx, userQuery)
  if err != nil || len(queryVector) == 0 {
    log.Printf("Embedding Error: %v", err)
    return "❌ Sorry, I had trouble processing your search. Please try again."
  }

  // 3. 🗄️ Fetch from DB
  relevantPosts, err := pkg.FetchRelevantPosts(db, queryVector, 5)
  if err != nil {
    log.Printf("DB Fetch Error: %v", err)
    return "❌ Database query failed. Please try again later."
  }

  if len(relevantPosts) == 0 {
    return "🔍 No matching items found. Try a different keyword!"
  }

  // 4. 🧠 LLM Analysis
  // Convert pkg.RedditPost to a generic map if your SearchUsedItems still expects map[string]interface{}
  // Or update SearchUsedItems to accept []pkg.RedditPost (Recommended)
  llmCtx, cancelLLM := context.WithTimeout(ctx, 30*time.Second)
  defer cancelLLM()

  // We convert the posts to a compatible slice for the LLM function
  postsMap := make([]map[string]interface{}, len(relevantPosts))
  for i, p := range relevantPosts {
    postsMap[i] = map[string]interface{}{
      "title":   p.Title,
      "content": p.Content,
      "price":   p.Price,
    }
  }

  result, err := SearchUsedItems(llmCtx, userQuery, postsMap)
  if err != nil || result == nil || len(result.Listings) == 0 {
    log.Printf("LLM Analysis Error: %v", err)
    return "❌ Failed to analyze results. Try a more specific search."
  }

  // 5. 🧾 Format final Telegram message
  reply := FormatTelegramMessage(userQuery, result, relevantPosts)
  
  return reply
}

